### DISCUSSION TITLE
"Certified" and "Unified" Community Packages - Is this Idea Realistic?

### DISCUSSION THREAD

**Discussion by msfredb7** [openupm/openupm#2446] [General]
Hi all! I've had this subject in mind for a while now and I wanted to get the community's opinion on it. I'm sort of new to community packages so most of you probably have a valuable insight that I don't have.

# TDLR
A trusted community-driven GitHub organisation with **robust and clean** packages that add essential features to the Unity engine and editor. Is this idea realistic? Has it been tried before? If so, has it failed and why?

# Full Version
Hi, I'm Fred! I've been using Unity for about 9 years now (professionally and personally). I work at Ubisoft on one of the projects that use Unity.

The engine is great, but many industry-standard tools and features are not implemented out of the box. Few examples:
* A game console, which allows you to use cheats and commands
* A color palette system
* Channels for the logging system
* Many scripts and editor utilities (e.g. Adding "Copy GUID" to an asset Right Click menu)
* etc. _(I'm sure you have many in mind)_

Many of these are developed in-house with custom solutions. I find it a bit sad that they are not part of the base engine because nobody else gets to use them.

### Why don't projects share and reuse more packages?
Like many of you, I dislike repeating the same code and I like sharing. So I find it strange that all the projects I've been on use very few external packages. Why isn't there more sharing and reuse? From my experience with external packages, most of them have one or many of the following deterrents:
* Package is a paid product (e.g. on the asset store)
* Package has poor discoverability (e.g. only way to find it is by knowing its strange name or a forum post)
* Package content is not trusted
    * _Are there game-breaking bugs? Is it performant? Does it compile on every platform? etc._
    * This is one of the most problematic issues from my experience. Most projects are too reluctant to try something unproven.
* Package owner or contributors are not trusted
    * _Will this package be updated to support newer versions of Unity? If I find bugs, will they be fixed? Are the contributors still active?_
    * This is the second biggest issue. Once projects trust a package's content, they need to be able to trust its support.

Let me know your opinion on this, I'd love to hear from your experiences.

### A Potential Solution
What if there was a trusted community organisation (e.g. a GitHub organisation) that hosted Unity packages with certified quality and support. I could imagine it working with guidelines such as:
* Everything is open source, no personal ownership
* All packages go through an approval process before being hosted. 
* All packages are test covered
* Rigorous code reviews and validations
* Unified code style
* Quality over quantity

Now, I don't even know if building an organisation like that is feasible. If it is, it's probably hard to get right, and very long to grow. 

### What's your take on this?
* Is this realistic? 
* Is the problem even real? 
* Has this been tried before.
    * If so, has it failed and why?

Most of the people here probably have a lot of insightful experiences that I lack. So let me know what you think!

**Comment by favoyang** [comment]
@msfredb7

Thanks for sharing your idea. OpenUPM is built to resolve part of the issue. But it only provides an aggregated way to organize open-source packages. Perhaps it is a "Unified" but not a "Certified" solution.

To achieve your proposal it requires a more central and strong ownership organization/community. Before building a group of certified or trusted packages you need a group of certified or trusted developers. It means more resources, It can be a challenge to build something like this without a foundation with good financial support. Even for OpenUPM itself, which built on the top of a donation model, with some sort of respect from the Unity community, but at the time of writing, it didn't receive enough funding to cover the operation cost, not to speak of hundreds of hours spent on developing and maintaining the service.

OpenUPM collects 1000+ UPM packages. Some of them as you said, are not well tested, quality not verified, no CR, various coding styles. Most developers start a project for their personal usage, then sharing with the community to make it more generic purposed. They may like to keep the ownership or won't have enough energy to keep improving the code quality. That's basically the way open-source works - Some open-source projects eventually survived to generate enough attention and support to keep their way growing. Well, most other open-source projects keep a loose relationship between developer and users, with a more few years of active development then goes to the long tail maintenance mode.

If you want to cover everything (with good quality in mind), but lacking good resources to execute, it may end with a failure. Even Unity has quite a few preview packages that they intended to end some unsuccessful experiments at some point. If you want to narrow down to a group of more specific features that connect to each other, you may eventually create a "framework". Like the way web development evolves a decade ago, developers design and pack useful, reusable components together and make called it Django, Rails, React... And it's hard to believe to start a web project without a good framework today.

Your idea remains me of TideLift, a subscription service to manage a group of tailored open-source packages with good tech support. It's a paid service, and I didn't know how it performs today. But it will never happen for Unity that the Package Terms prevent such commercial practice and Unity Asset Store is their option for paid content.

For OpenUPM, we actually have many points to improve to help developers find packages with good reputations. But considering our limited impaction and resources, it will be a slow pace.

**Reply by msfredb7** [reply]
Thanks for your extensive reply @favoyang.

As you said, I think one of the first keys to success for this sort of project is a group of trusted developers. That may only be kickstart-able with a big financial backing.

The end goal is that, if enough people use the packages and contribute then, maintenance will naturally work on its own without needing much more than a few donations to cover the github fees. But maybe it's a bit optimistic. 

Anyway, thanks again foe sharing your insight. I didn't know about the history of web frameworks and TideLift
